irohad/ordering/impl/ordering_gate_impl.cpp

```diff
@@ -51,27 +58,60 @@ namespace iroha {
 
     void OrderingGateImpl::setPcs(
         const iroha::network::PeerCommunicationService &pcs) {
-      pcs_subscriber_ = pcs.on_commit().subscribe([this](auto) {
-        // TODO: 05/03/2018 @muratovv rework behavior of queue with respect to
-        // block height IR-1042
-        unlock_next_.store(true);
-        this->tryNextRound();
+      log_->info("setPcs");
 
-      });
+      auto top_block_height = pcs.on_commit()
+                                  .transform([](const Commit &block) {
+                                    // find height of last commited block
+                                    return block.as_blocking().last()->height();
+                                  })
+                                  .start_with(last_block_height_);
+
+      auto subscribe = [&](auto merge_strategy) {
+        pcs_subscriber_ = merge_strategy(net_proposals_.get_observable())
+                              .subscribe([this](const auto &t) {
+                                this->tryNextRound(std::get<1>(t));
+                              });
+      };
+
+      if (run_async_) {
+       subscribe([&top_block_height](auto observable) {
+         return observable.combine_latest(rxcpp::synchronize_new_thread(),
+                         top_block_height);
+       });
+      } else {
+        subscribe([&top_block_height](auto observable) {
+          return observable.combine_latest(top_block_height);
+        });
+      }
     }
 
     void OrderingGateImpl::onProposal(
         std::shared_ptr<shared_model::interface::Proposal> proposal) {
-      log_->info("Received new proposal");
+      log_->info("Received new proposal, height: {}", proposal->height());
       proposal_queue_.push(std::move(proposal));
-      tryNextRound();
+      std::lock_guard<std::mutex> lock(proposal_mutex_);
+      net_proposals_.get_subscriber().on_next(0);
     }
 
-    void OrderingGateImpl::tryNextRound() {
-      if (not proposal_queue_.empty() and unlock_next_.exchange(false)) {
-        std::shared_ptr<shared_model::interface::Proposal> next_proposal;
-        proposal_queue_.try_pop(next_proposal);
-        log_->info("Pass the proposal to pipeline");
+    void OrderingGateImpl::tryNextRound(
+        shared_model::interface::types::HeightType last_block_height_) {
+      log_->info("TryNextRound");
+      std::shared_ptr<shared_model::interface::Proposal> next_proposal;
+      while (proposal_queue_.try_pop(next_proposal)) {
+        // check for old proposal
+        if (next_proposal->height() < last_block_height_ + 1) {
```

Comment:
`(next_proposal->height() <= last_block_height_)`?

Reply:
I think it is more clear that we need proposal which is one older than last commited block